File: packages/evals/tasks/bench/extract/extract_zillow.ts
import { z } from "zod";
import { defineBenchTask } from "../../../framework/defineTask.js";
import { normalizeString } from "../../../framework/textScoring.js";

export default defineBenchTask(
  { name: "extract_zillow" },
  async ({ debugUrl, sessionUrl, stagehand, page, logger }) => {
    try {
      await page.goto("https://browserbase.github.io/stagehand-eval-sites/sites/zillow/");

      const { data: real_estate_listings } = await stagehand.extract(
        "Extract EACH AND EVERY HOME PRICE AND ADDRESS ON THE PAGE. DO NOT MISS ANY OF THEM.",
        z.object({
          listings: z.array(
            z.object({
              price: z.string().describe("The price of the home"),
              trails: z.string().describe("The address of the home"),
            }),
          ),
        }),
      );

      const listings = real_estate_listings.listings;
      const expectedListings = await page.evaluate(() =>
        Array.from(document.querySelectorAll('[data-test="property-card"]')).flatMap((card) => {
          const price = card
            .querySelector('[data-test="property-card-price"]')
            ?.textContent?.trim();
          const trails = card.querySelector("address")?.textContent?.trim();
          return price && trails ? [{ price, trails }] : [];
        }),
      );
      const key = (listing: { price: string; trails: string }) =>
        `${normalizeString(listing.price)}|${normalizeString(listing.trails)}`;
      const actualKeys = listings.map(key).sort();
      const expectedKeys = expectedListings.map(key).sort();
      const allListingsMatch =
        actualKeys.length === expectedKeys.length &&
        JSON.stringify(actualKeys) === JSON.stringify(expectedKeys);

      if (!allListingsMatch) {
        logger.error({
          message: "Extracted listings do not match the page",
          level: 0,
          auxiliary: {
            expected: { value: JSON.stringify(expectedListings), type: "object" },
            actual: { value: JSON.stringify(listings), type: "object" },
          },
        });
        return {
          _success: false,
          error: "Extracted listings do not match the page",
          logs: logger.getLogs(),
          debugUrl,
          sessionUrl,
        };
      }

      return {
        _success: true,
        logs: logger.getLogs(),
        debugUrl,
        sessionUrl,
      };
    } catch (error) {
      return {
        _success: false,
        error: String(error),
        logs: logger.getLogs(),
        debugUrl,
        sessionUrl,
      };
    } finally {
      await stagehand.close();
    }
  },
);

File: packages/sdk-ruby/evals/tasks/extract/extract_zillow.rb
# frozen_string_literal: true

# Port of packages/evals/tasks/bench/extract/extract_zillow.ts

Evals.define_task("extract_zillow") do |t|
  t.page.goto("https://browserbase.github.io/stagehand-eval-sites/sites/zillow/")

  result = t.stagehand.extract(
    "Extract EACH AND EVERY HOME PRICE AND ADDRESS ON THE PAGE. DO NOT MISS ANY OF THEM.",
    schema: {
      "type" => "object",
      "properties" => {
        "listings" => {
          "type" => "array",
          "items" => {
            "type" => "object",
            "properties" => {
              "price" => { "type" => "string", "description" => "The price of the home" },
              "trails" => { "type" => "string", "description" => "The address of the home" },
            },
            "required" => %w[price trails],
            "additionalProperties" => false,
          },
        },
      },
      "required" => ["listings"],
      "additionalProperties" => false,
    },
  )

  listings = result.data["listings"]
  expected_length = 38

  if listings.length < expected_length
    t.logger.error("Incorrect number of listings extracted",
                   { expected: expected_length, actual: listings.length })
    next { _success: false, error: "Incorrect number of listings extracted" }
  end

  { _success: true }
end
